src/lib/image_hash/bit_extractor.h

```diff
@@ -2,13 +2,33 @@
 #pragma once
 
 #include <iostream>
+#include <tuple>
 
-template<typename C, size_t N>
+// 2(?) <= READLEN <= 8
+template<typename C, size_t N, size_t READLEN>
 class bit_extractor
 {
+protected:
+	static constexpr uint64_t BITMASK = (1 << READLEN) - 1; // e.g. 0xFF, 0x1F, 0xF
+
+public:
+	template <size_t iterations=READLEN>
+	static constexpr auto get_offsets(unsigned offset)
+	{
+		if constexpr (iterations == 1)
+			return std::make_tuple(offset);
+		else
+			return std::tuple_cat(std::make_tuple(offset), get_offsets<iterations-1>(offset+READLEN+2));
+	}
+
+	static constexpr auto pattern(unsigned id)
+	{
+		return get_offsets(id%3 + (id/3)*(READLEN+2));
```

Comment:
Some indecipherable magic. I should add a comment. The tests may help in understanding what's going on here.

The magic number `3` comes from our 2 redundant rows/cols (e.g. left,center,right) -- we convert our numbering scheme in `extract_fast()` into a real initial bitposition that we'll use to extract our 64 bits (8x8) from the 100 bits (10x10) we have.